Spigot-Server-Patches/0409-Add-player-view-distance-API.patch

```diff
@@ -0,0 +1,339 @@
++        this.viewDistance = viewDistance;
++    }
+     // Paper end
+ 
+     // CraftBukkit start
+diff --git a/src/main/java/net/minecraft/server/EntityWither.java b/src/main/java/net/minecraft/server/EntityWither.java
+index e3feb39ad..bc6f19ff0 100644
+--- a/src/main/java/net/minecraft/server/EntityWither.java
++++ b/src/main/java/net/minecraft/server/EntityWither.java
+@@ -210,7 +210,7 @@ public class EntityWither extends EntityMonster implements IRangedEntity {
+                 // Paper start
+                 //int viewDistance = ((WorldServer) this.world).spigotConfig.viewDistance * 16; // Paper - updated to use worlds actual view distance incase we have to uncomment this due to removal of player view distance API
+                 for (EntityPlayer player : ((WorldServer)world).getPlayers()) {
+-                    final int viewDistance = player.getViewDistance(); // TODO apply view distance api patch
++                    final int viewDistance = player.getViewDistance(); // Paper - apply view distance api patch
+                     // Paper end
+                     double deltaX = this.locX - player.locX;
+                     double deltaZ = this.locZ - player.locZ;
+diff --git a/src/main/java/net/minecraft/server/PlayerChunkMap.java b/src/main/java/net/minecraft/server/PlayerChunkMap.java
+index fba1639a8..07a424b68 100644
+--- a/src/main/java/net/minecraft/server/PlayerChunkMap.java
++++ b/src/main/java/net/minecraft/server/PlayerChunkMap.java
+@@ -743,6 +743,7 @@ public class PlayerChunkMap extends IChunkLoader implements PlayerChunk.d {
+         }
+     }
+ 
++    public int getViewDistance() { return this.viewDistance; } // Paper
+     protected void setViewDistance(int i, int j) {
+         int k = MathHelper.clamp(i + 1, 3, 33);
+         int l;
+@@ -753,20 +754,25 @@ public class PlayerChunkMap extends IChunkLoader implements PlayerChunk.d {
+             this.u.b(this.viewDistance);
+             ObjectIterator objectiterator = this.updatingChunks.values().iterator();
+ 
+-            while (objectiterator.hasNext()) {
+-                PlayerChunk playerchunk = (PlayerChunk) objectiterator.next();
+-                ChunkCoordIntPair chunkcoordintpair = playerchunk.h();
+-                Packet<?>[] apacket = new Packet[2];
+-
+-                int finall = l; // CraftBukkit - decompile error
+-                this.a(chunkcoordintpair, false).forEach((entityplayer) -> {
+-                    int i1 = b(chunkcoordintpair, entityplayer, true);
+-                    boolean flag = i1 <= finall; // CraftBukkit - decompile error
+-                    boolean flag1 = i1 <= this.viewDistance;
+-
+-                    this.sendChunk(entityplayer, chunkcoordintpair, apacket, flag, flag1);
+-                });
++            // Paper start - Separate into two methods
++            if(this.world.players != null) {
++                new ArrayList<>(this.world.players).forEach((player) -> this.setViewDistance(player, k));
```

Comment:
Why `new ArrayList` ?

Reply:
i think that can allow use lambda to foreach

Reply:
Or you can use `Arrays.stream(this.world.players)`, so Java doesn't need to allocate new memory 😄